File: 2020/day_10/second_part.py
from typing import List

adapters: List[int] = [0]
adapter = int(input())
while adapter != -1:
    adapters.append(adapter)
    adapter = int(input())
adapters.append(max(adapters) + 3)
adapters.sort()

nb_arrangements = [0] * len(adapters)
nb_arrangements[-1] = 1

for i in range(len(adapters)-1, -1, -1):
    for j in (1,2,3,):
        if i-j >= 0 and adapters[i-j] + 3 >= adapters[i]:
            nb_arrangements[i-j] += nb_arrangements[i]

print("solutions:", nb_arrangements[0])

File: 2020/day_10/haskell/second_part.hs
module Main where

import Data.List
import System.IO

main :: IO ()
main = do
  m <- doReadFile
  print $ 
    nbArrangements 
      (reverse $ sort $ map read m :: [Int])
      []
  return ()

nbArrangements :: [Int] -> [Int] -> [Int]
countDiff js memo 
  | null js = memo 
  | otherwise = case (head js) -  of
    1 -> 
      
      
addTuple :: Num a => (a, a, a) -> (a, a, a) -> (a, a, a)
addTuple (a1, b1, c1) (a2, b2, c2) = (a1+a2, b1+b2, c1+c2)

doReadFile :: IO [[Char]]
doReadFile = withFile "input" ReadMode doReadLine

doReadLine :: Handle -> IO [[Char]]
doReadLine hFile = do
  isEnd <- hIsEOF hFile
  if isEnd
    then return []
    else do
      line <- hGetLine hFile
      rest <- doReadLine hFile
      return (line : rest)